problems building on os X stddef.h and math.h not found

I'm using a M1 mac with most packages installed via brew (including python).

I'm trying to follow the cmake instructions from https://warpx.readthedocs.io/en/latest/install/users.html.

I tried this both on the 24.08 tag and the latest git commit (3dda26f)

Creating the build files seems to work:
```
arun@Aruns-MacBook-Pro-2:~/src/WarpX((24.08))>cmake -S . -B build
-- The C compiler identification is AppleClang 15.0.0.15000309
-- The CXX compiler identification is AppleClang 15.0.0.15000309
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /Library/Developer/CommandLineTools/usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /Library/Developer/CommandLineTools/usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Could NOT find CCache
-- Downloading AMReX ...
-- AMReX repository: https://github.com/AMReX-Codes/amrex.git (24.08)
CMake Warning (dev) at /opt/homebrew/Cellar/cmake/3.30.2/share/cmake/Modules/FetchContent.cmake:1953 (message):
  Calling FetchContent_Populate(fetchedamrex) is deprecated, call
  FetchContent_MakeAvailable(fetchedamrex) instead.  Policy CMP0169 can be
  set to OLD to allow FetchContent_Populate(fetchedamrex) to be called
  directly for now, but the ability to call it with declared details will be
  removed completely in a future version.
Call Stack (most recent call first):
  cmake/dependencies/AMReX.cmake:151 (FetchContent_Populate)
  cmake/dependencies/AMReX.cmake:280 (find_amrex)
  CMakeLists.txt:209 (include)
This warning is for project developers.  Use -Wno-dev to suppress it.

-- CMake version: 3.30.2
-- AMReX installation directory: /usr/local
-- Build type set by user to 'Release'.
-- Building AMReX with AMReX_SPACEDIM = 3
-- Configuring AMReX with the following options enabled: 
--    AMReX_PRECISION = DOUBLE
--    AMReX_MPI
--    AMReX_MPI_THREAD_MULTIPLE
--    AMReX_OMP
--    AMReX_LINEAR_SOLVERS
--    AMReX_PARTICLES
--    AMReX_PARTICLES_PRECISION = DOUBLE
--    AMReX_PIC
--    AMReX_EXPORT_DYNAMIC
--    AMReX_TINY_PROFILE
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE
-- Found MPI_C: /opt/homebrew/Cellar/open-mpi/5.0.3_1/lib/libmpi.dylib (found version "3.1")
-- Found MPI_CXX: /opt/homebrew/Cellar/open-mpi/5.0.3_1/lib/libmpi.dylib (found version "3.1")
-- Found MPI: TRUE (found version "3.1") found components: C CXX
-- Found OpenMP_CXX: -Xclang -fopenmp (found version "5.0")
-- Found OpenMP: TRUE (found version "5.0") found components: CXX
-- AMReX configuration summary: 
--    Build type               = Release
--    Install directory        = /usr/local
--    C++ compiler             = /Library/Developer/CommandLineTools/usr/bin/c++
--    C++ defines              = -DAMREX_SPACEDIM=3
--    C++ flags                = -O3 -DNDEBUG  -Xclang -fopenmp
--    C++ include paths        = -I/Users/arun/src/WarpX/build/_deps/fetchedamrex-src/Src/Base -I/Users/arun/src/WarpX/build/_deps/fetchedamrex-src/Src/Base/Parser -I/Users/arun/src/WarpX/build/_deps/fetchedamrex-src/Src/Boundary -I/Users/arun/src/WarpX/build/_deps/fetchedamrex-src/Src/AmrCore -I/Users/arun/src/WarpX/build/_deps/fetchedamrex-src/Src/LinearSolvers -I/Users/arun/src/WarpX/build/_deps/fetchedamrex-src/Src/LinearSolvers/MLMG -I/Users/arun/src/WarpX/build/_deps/fetchedamrex-src/Src/LinearSolvers/OpenBC -I/Users/arun/src/WarpX/build/_deps/fetchedamrex-src/Src/Particle -I/opt/homebrew/Cellar/open-mpi/5.0.3_1/include -I/opt/homebrew/include
--    Link line                = /opt/homebrew/Cellar/open-mpi/5.0.3_1/lib/libmpi.dylib /opt/homebrew/lib/libomp.dylib
-- AMReX: Using version '24.08' (24.08)
-- Downloading PICSAR ...
-- PICSAR repository: https://github.com/ECP-WarpX/picsar.git (aa54e985398c1d575abc7e6737cdbc660a13765f)
CMake Warning (dev) at /opt/homebrew/Cellar/cmake/3.30.2/share/cmake/Modules/FetchContent.cmake:1953 (message):
  Calling FetchContent_Populate(fetchedpicsar) is deprecated, call
  FetchContent_MakeAvailable(fetchedpicsar) instead.  Policy CMP0169 can be
  set to OLD to allow FetchContent_Populate(fetchedpicsar) to be called
  directly for now, but the ability to call it with declared details will be
  removed completely in a future version.
Call Stack (most recent call first):
  cmake/dependencies/PICSAR.cmake:63 (FetchContent_Populate)
  cmake/dependencies/PICSAR.cmake:125 (find_picsar)
  CMakeLists.txt:219 (include)
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Downloading openPMD-api ...
-- openPMD-api repository: https://github.com/openPMD/openPMD-api.git (0.15.2)
CMake Warning (dev) at /opt/homebrew/Cellar/cmake/3.30.2/share/cmake/Modules/FetchContent.cmake:1953 (message):
  Calling FetchContent_Populate(fetchedopenpmd) is deprecated, call
  FetchContent_MakeAvailable(fetchedopenpmd) instead.  Policy CMP0169 can be
  set to OLD to allow FetchContent_Populate(fetchedopenpmd) to be called
  directly for now, but the ability to call it with declared details will be
  removed completely in a future version.
Call Stack (most recent call first):
  cmake/dependencies/openPMD.cmake:38 (FetchContent_Populate)
  cmake/dependencies/openPMD.cmake:100 (find_openpmd)
  CMakeLists.txt:223 (include)
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Found MPI: TRUE (found version "3.1") found components: CXX
-- Using the single-header code from /Users/arun/src/WarpX/build/_deps/fetchedopenpmd-src/share/openPMD/thirdParty/json/single_include/
-- nlohmann-json: Using INTERNAL version '3.9.1'
-- toml11: Using INTERNAL version '3.7.1'
-- Found HDF5: /opt/homebrew/Cellar/hdf5-mpi/1.14.3/lib/libhdf5.dylib;/opt/homebrew/opt/libaec/lib/libsz.dylib;/Library/Developer/CommandLineTools/SDKs/MacOSX14.4.sdk/usr/lib/libz.tbd;/Library/Developer/CommandLineTools/SDKs/MacOSX14.4.sdk/usr/lib/libdl.tbd;/Library/Developer/CommandLineTools/SDKs/MacOSX14.4.sdk/usr/lib/libm.tbd (found version "1.14.3") found components: C
-- Found MPI: TRUE (found version "3.1")
-- Found ADIOS2: /opt/homebrew/lib/cmake/adios2/adios2-config.cmake (found suitable version "2.10.1", minimum required is "2.7.0") found components: C CXX MPI

openPMD build configuration:
  library Version: 0.15.2
  openPMD Standard: 1.1.0
  C++ Compiler: AppleClang 15.0.0.15000309 
    /Library/Developer/CommandLineTools/usr/bin/c++

  Installation: OFF

  Build Type: Release
  Library: static
  CLI Tools: OFF
  Examples: OFF
  Testing: OFF
  Invasive Tests: OFF
  Internal VERIFY: ON
  Build Options:
    MPI: ON
    HDF5: ON
    ADIOS1: OFF
    ADIOS2: ON
    PYTHON: OFF
    CUDA_EXAMPLES: OFF

-- Found Git: /opt/homebrew/bin/git (found version "2.46.0")
-- Performing Test COMPILER_HAS_HIDDEN_VISIBILITY
-- Performing Test COMPILER_HAS_HIDDEN_VISIBILITY - Success
-- Performing Test COMPILER_HAS_HIDDEN_INLINE_VISIBILITY
-- Performing Test COMPILER_HAS_HIDDEN_INLINE_VISIBILITY - Success
-- Performing Test COMPILER_HAS_DEPRECATED_ATTR
-- Performing Test COMPILER_HAS_DEPRECATED_ATTR - Success

WarpX build configuration:
  Version: 24.08 (24.08)
  C++ Compiler: AppleClang 15.0.0.15000309 
    /Library/Developer/CommandLineTools/usr/bin/c++

  Installation prefix: /usr/local
        bin: bin
        lib: lib
    include: include
      cmake: lib/cmake/WarpX

  Build type: Release
  Build options:
    APP: ON
    ASCENT: OFF
    COMPUTE: OMP
    DIMS: 3
    Embedded Boundary: OFF
    IPO/LTO: OFF
    LIB: ON (static)
    MPI: ON
    PARTICLE PRECISION: DOUBLE
    PRECISION: DOUBLE
    FFT Solvers: OFF
    heFFTe: OFF
    PYTHON: OFF
    OPENPMD: ON
    QED: ON
    QED table generation: OFF
    QED tools: OFF
    SENSEI: OFF

-- Configuring done (13.8s)
-- Generating done (0.2s)
-- Build files have been written to: /Users/arun/src/WarpX/build

```

But when I try to compile it, I run into errors:
```

arun@Aruns-MacBook-Pro-2:~/src/WarpX((24.08))>cmake --build build -j 4
[  0%] Building CXX object _deps/fetchedopenpmd-build/CMakeFiles/openPMD.dir/src/config.cpp.o
[  1%] Building CXX object _deps/fetchedopenpmd-build/CMakeFiles/openPMD.dir/src/ChunkInfo.cpp.o
[  1%] Building CXX object CMakeFiles/buildInfoapp_3d.dir/app_3d/AMReX_buildInfo.cpp.o
[  1%] Building CXX object _deps/fetchedamrex-build/Src/CMakeFiles/amrex_3d.dir/Base/AMReX_BlockMutex.cpp.o
In file included from /Users/arun/src/WarpX/build/_deps/fetchedopenpmd-src/src/ChunkInfo.cpp:21:
In file included from /Users/arun/src/WarpX/build/_deps/fetchedopenpmd-src/include/openPMD/ChunkInfo.hpp:23:
In file included from /Users/arun/src/WarpX/build/_deps/fetchedopenpmd-src/include/openPMD/Dataset.hpp:23:
In file included from /Users/arun/src/WarpX/build/_deps/fetchedopenpmd-src/include/openPMD/Datatype.hpp:23:
In file included from /Users/arun/src/WarpX/build/_deps/fetchedopenpmd-src/include/openPMD/auxiliary/TypeTraits.hpp:24:
In file included from /Users/arun/src/WarpX/build/_deps/fetchedopenpmd-src/include/openPMD/auxiliary/UniquePtr.hpp:3:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX14.4.sdk/usr/include/c++/v1/functional:515:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX14.4.sdk/usr/include/c++/v1/__algorithm/search.h:14:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX14.4.sdk/usr/include/c++/v1/__algorithm/iterator_operations.h:12:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX14.4.sdk/usr/include/c++/v1/__algorithm/iter_swap.h:14:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX14.4.sdk/usr/include/c++/v1/__utility/swap.h:16:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX14.4.sdk/usr/include/c++/v1/__type_traits/is_nothrow_move_constructible.h:15:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX14.4.sdk/usr/include/c++/v1/__type_traits/is_nothrow_constructible.h:17:
/Library/Developer/CommandLineTools/SDKs/MacOSX14.4.sdk/usr/include/c++/v1/cstddef:46:5: error: <cstddef> tried including <stddef.h> but didn't find libc++'s <stddef.h> header.           This usually means that your header search paths are not configured properly.           The header search paths should contain the C++ Standard Library headers before           any C Standard Library, and you are probably using compiler flags that make that           not be the case.
#   error <cstddef> tried including <stddef.h> but didn't find libc++'s <stddef.h> header. \
    ^
In file included from /Users/arun/src/WarpX/build/_deps/fetchedopenpmd-src/src/config.cpp:22:
In file included from /Users/arun/src/WarpX/build/_deps/fetchedopenpmd-src/include/openPMD/version.hpp:23:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX14.4.sdk/usr/include/c++/v1/map:574:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX14.4.sdk/usr/include/c++/v1/__algorithm/equal.h:14:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX14.4.sdk/usr/include/c++/v1/__algorithm/unwrap_iter.h:13:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX14.4.sdk/usr/include/c++/v1/__iterator/iterator_traits.h:14:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX14.4.sdk/usr/include/c++/v1/__concepts/constructible.h:12:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX14.4.sdk/usr/include/c++/v1/__concepts/convertible_to.h:13:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX14.4.sdk/usr/include/c++/v1/__type_traits/is_convertible.h:14:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX14.4.sdk/usr/include/c++/v1/__type_traits/is_array.h:14:
/Library/Developer/CommandLineTools/SDKs/MacOSX14.4.sdk/usr/include/c++/v1/cstddef:46:5: error: <cstddef> tried including <stddef.h> but didn't find libc++'s <stddef.h> header.           This usually means that your header search paths are not configured properly.           The header search paths should contain the C++ Standard Library headers before           any C Standard Library, and you are probably using compiler flags that make that           not be the case.
#   error <cstddef> tried including <stddef.h> but didn't find libc++'s <stddef.h> header. \
    ^
/Library/Developer/CommandLineTools/SDKs/MacOSX14.4.sdk/usr/include/c++/v1/cstddef:59:9: error: no member named 'nullptr_t' in the global namespace
using ::nullptr_t;
      ~~^
/Library/Developer/CommandLineTools/SDKs/MacOSX14.4.sdk/usr/include/c++/v1/cstddef:59:9: error: no member named 'nullptr_t' in the global namespace
using ::nullptr_t;
      ~~^
In file included from /Users/arun/src/WarpX/build/_deps/fetchedamrex-src/Src/Base/AMReX_BlockMutex.cpp:1:
In file included from /Users/arun/src/WarpX/build/_deps/fetchedamrex-src/Src/Base/AMReX_BlockMutex.H:5:
In file included from /Users/arun/src/WarpX/build/_deps/fetchedamrex-src/Src/Base/AMReX_Gpu.H:17:
In file included from /Users/arun/src/WarpX/build/_deps/fetchedamrex-src/Src/Base/AMReX_GpuControl.H:8:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX14.4.sdk/usr/include/c++/v1/utility:251:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX14.4.sdk/usr/include/c++/v1/__utility/as_const.h:14:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX14.4.sdk/usr/include/c++/v1/__utility/forward.h:15:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX14.4.sdk/usr/include/c++/v1/__type_traits/remove_reference.h:13:
/Library/Developer/CommandLineTools/SDKs/MacOSX14.4.sdk/usr/include/c++/v1/cstddef:46:5: error: <cstddef> tried including <stddef.h> but didn't find libc++'s <stddef.h> header.           This usually means that your header search paths are not configured properly.           The header search paths should contain the C++ Standard Library headers before           any C Standard Library, and you are probably using compiler flags that make that           not be the case.
#   error <cstddef> tried including <stddef.h> but didn't find libc++'s <stddef.h> header. \
    ^
/Library/Developer/CommandLineTools/SDKs/MacOSX14.4.sdk/usr/include/c++/v1/cstddef:59:9: error: no member named 'nullptr_t' in the global namespace
using ::nullptr_t;
      ~~^
In file included from /Users/arun/src/WarpX/build/_deps/fetchedopenpmd-src/src/config.cpp:22:
In file included from /Users/arun/src/WarpX/build/_deps/fetchedopenpmd-src/include/openPMD/version.hpp:23:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX14.4.sdk/usr/include/c++/v1/map:574:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX14.4.sdk/usr/include/c++/v1/__algorithm/equal.h:14:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX14.4.sdk/usr/include/c++/v1/__algorithm/unwrap_iter.h:13:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX14.4.sdk/usr/include/c++/v1/__iterator/iterator_traits.h:14:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX14.4.sdk/usr/include/c++/v1/__concepts/constructible.h:13:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX14.4.sdk/usr/include/c++/v1/__concepts/destructible.h:13:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX14.4.sdk/usr/include/c++/v1/__type_traits/is_nothrow_destructible.h:16:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX14.4.sdk/usr/include/c++/v1/__type_traits/is_scalar.h:17:
/Library/Developer/CommandLineTools/SDKs/MacOSX14.4.sdk/usr/include/c++/v1/__type_traits/is_null_pointer.h:26:28: error: use of undeclared identifier 'nullptr_t'; did you mean 'nullptr'?
struct __is_nullptr_t_impl<nullptr_t> : public true_type {};
                           ^
/Library/Developer/CommandLineTools/SDKs/MacOSX14.4.sdk/usr/include/c++/v1/__type_traits/is_null_pointer.h:26:28: error: template argument for template type parameter must be a type
struct __is_nullptr_t_impl<nullptr_t> : public true_type {};
                           ^~~~~~~~~
/Library/Developer/CommandLineTools/SDKs/MacOSX14.4.sdk/usr/include/c++/v1/__type_traits/is_null_pointer.h:23:17: note: template parameter is declared here
template <class _Tp>

```

I also see errors in regards to <stdint.h>, <stdlib.h>, and <math.h>. 

Is this a problem when creating the build scripts or just some issue with my CC-compiler not knowing the path to these files (which seems odd to me)?

Let me know if I should upload any more files or logs.

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -254,7 +254,7 @@ endif()
 
 # Python
 if(WarpX_PYTHON)
-    find_package(Python COMPONENTS Interpreter Development.Module REQUIRED)
+    find_package(Python 3.8 COMPONENTS Interpreter Development.Module REQUIRED)
 
     # default installation directories: Python
     warpx_set_default_install_dirs_python()

--- a/cmake/dependencies/pybind11.cmake
+++ b/cmake/dependencies/pybind11.cmake
@@ -10,6 +10,11 @@ function(find_pybind11)
         message(STATUS "pybind11 repository: ${WarpX_pybind11_repo} (${WarpX_pybind11_branch})")
         include(FetchContent)
     endif()
+
+    # rely on our find_package(Python ...) call
+    # https://pybind11.readthedocs.io/en/stable/compiling.html#modules-with-cmake
+    set(PYBIND11_FINDPYTHON ON)
+
     if(WarpX_pybind11_internal OR WarpX_pybind11_src)
         set(CMAKE_POLICY_DEFAULT_CMP0077 NEW)
 

--- a/setup.py
+++ b/setup.py
@@ -84,7 +84,15 @@ class CMakeBuild(build_ext):
         r_dim = re.search(r"warpx_(1|2|rz|3)(?:d*)", ext.name)
         dims = r_dim.group(1).upper()
 
+        pyv = sys.version_info
         cmake_args = [
+            # Python: use the calling interpreter in CMake
+            # https://cmake.org/cmake/help/latest/module/FindPython.html#hints
+            # https://cmake.org/cmake/help/latest/command/find_package.html#config-mode-version-selection
+            f"-DPython_ROOT_DIR={sys.prefix}",
+            f"-DPython_FIND_VERSION={pyv.major}.{pyv.minor}.{pyv.micro}",
+            "-DPython_FIND_VERSION_EXACT=TRUE",
+            "-DPython_FIND_STRATEGY=LOCATION",
             "-DCMAKE_LIBRARY_OUTPUT_DIRECTORY=" + os.path.join(extdir, "pywarpx"),
             "-DCMAKE_RUNTIME_OUTPUT_DIRECTORY=" + extdir,
             "-DWarpX_DIMS=" + dims,

